store/nosqlstore/webhook_store.go

```diff
@@ -0,0 +1,347 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package nosqlstore
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/store"
+)
+
+type WebhookStore struct {
+	driver           Driver
+	incomingWebhooks ObjectStore
+	outgoingWebhooks ObjectStore
+}
+
+func NewWebhookStore(driver Driver) (*WebhookStore, error) {
+	s := &WebhookStore{
+		driver: driver,
+		incomingWebhooks: driver.ObjectStore("incoming_webhooks", json.Marshal, func(b []byte) (interface{}, error) {
+			var wh model.IncomingWebhook
+			return &wh, json.Unmarshal(b, &wh)
+		}),
+		outgoingWebhooks: driver.ObjectStore("outgoing_webhooks", json.Marshal, func(b []byte) (interface{}, error) {
+			var wh model.OutgoingWebhook
+			return &wh, json.Unmarshal(b, &wh)
+		}),
+	}
+
+	for name, f := range map[string]func(wh *model.IncomingWebhook) ([]byte, []byte){
+		"delete_at": func(wh *model.IncomingWebhook) ([]byte, []byte) { return Encode(wh.DeleteAt), []byte(wh.Id) },
+		"user_id":   func(wh *model.IncomingWebhook) ([]byte, []byte) { return []byte(wh.UserId), Encode(wh.DeleteAt, wh.Id) },
+		"channel_id": func(wh *model.IncomingWebhook) ([]byte, []byte) {
+			return []byte(wh.ChannelId), Encode(wh.DeleteAt, wh.Id)
+		},
+		"team_id": func(wh *model.IncomingWebhook) ([]byte, []byte) { return []byte(wh.TeamId), Encode(wh.DeleteAt, wh.Id) },
```

Comment:
The messiest part of all this is probably going to be the indexing. For the most part, with SQL we don't need to know what we're going to be querying ahead of time; any arbitrary query can be executed with no table modifications. With NoSQL, we have to prepare ahead of time with lookup tables and indexes.

Reply:
Can you elaborate on this requirement? My experience with NoSQL is mostly limited to MongoDB, and while you'd never want to do a table scan, it would still work. Won't a relational database do a table scan if it can't find a suitable index too?

Or is this comment a specific restriction of NoSQL databases like DynamoDB?

Reply:
Sorry, I guess we don't *have to* prepare ahead of time. We just really want to. Of course full scans are still possible.